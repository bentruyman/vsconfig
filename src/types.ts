import type { Preset } from "./presets";

interface Folder {
  name?: string;
  path: string;
}

export interface VsConfig {
  folders: Folder[];
  settings?: Preset["settings"];
}
