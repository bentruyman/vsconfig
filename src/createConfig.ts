import merge from "lodash.merge";
import type { PresetMap } from "./presets";

interface Folder {
  name?: string;
  path: string;
}
export type Settings = Record<string, unknown>;

export interface VsConfig {
  folders: Folder[];
  settings?: Settings;
}

export interface ConfigContext {
  presets: PresetMap;
}

export function createConfig(ctx: ConfigContext): VsConfig {
  const config: VsConfig = { folders: [{ path: "." }] };

  ctx.presets.forEach((preset) => {
    const presetSettings =
      typeof preset.settings === "function"
        ? preset.settings(ctx)
        : preset.settings;

    config.settings = merge(config.settings, presetSettings);
  });

  return config;
}
