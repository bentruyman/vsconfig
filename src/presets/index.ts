import deno from "./deno";
import javascript from "./javascript";
import javascriptreact from "./javascriptreact";
import json from "./json";
import markdown from "./markdown";
import typescript from "./typescript";
import typescriptreact from "./typescriptreact";
import yaml from "./yaml";

export interface Preset {
  settings: any;
}

export const presets: Map<string, Preset> = new Map([
  ["deno", deno],
  ["javascript", javascript],
  ["javascriptreact", javascriptreact],
  ["json", json],
  ["markdown", markdown],
  ["typescript", typescript],
  ["typescriptreact", typescriptreact],
  ["yaml", yaml],
]);
