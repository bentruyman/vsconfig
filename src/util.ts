import { basename, join } from "path";
import type { PresetMap } from "./presets";

const CONFIG_EXT = ".code-workspace";

export function createConfigFilename(cwd: string): string {
  return join(cwd, `${basename(cwd)}${CONFIG_EXT}`);
}

export function findMatchingPresets(
  filename: string,
  presets: PresetMap
): string[] {
  const matches: string[] = [];

  presets.forEach((value, key) => {
    if (filename.match(value.matcher)) {
      matches.push(key);
    }
  });

  return matches;
}
