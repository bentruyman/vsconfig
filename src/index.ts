import merge from "lodash.merge";
import { presets } from "./presets";
import type { VsConfig } from "./types";

interface CreateConfigOptions {
  presets: string[];
}

export function createConfig(options: CreateConfigOptions): VsConfig {
  const config: VsConfig = { folders: [{ path: "." }] };

  options.presets.forEach((name) => {
    const preset = presets.get(name);

    if (preset !== undefined) {
      config.settings = merge(config.settings, preset.settings);
    }
  });

  return config;
}
