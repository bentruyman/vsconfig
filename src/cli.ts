import cac from "cac";
import { writeFile } from "fs/promises";
import inquirer from "inquirer";
import { createConfig, scan } from ".";

import { presets } from "./presets";
import { createConfigFilename } from "./util";

interface CliArgs {
  exclude: string | string[];
  scan: boolean;
}

interface Answers {
  presets: string[];
}

const EXCLUDE_BASE = [".git", "node_modules"];

const cli = cac();

cli
  .command("[dir]", "Generate a new VSCode workspace configuration")
  .option(
    "-e, --exclude <...pattern>",
    "Exclude files during scan (only used with --scan)",
    { default: EXCLUDE_BASE }
  )
  .option(
    "--scan",
    "Scan the current project to generate a list of recommended presets",
    { default: false }
  )
  .action(
    async (dir = process.cwd(), { exclude, scan: shouldScan }: CliArgs) => {
      const initialPresets: string[] = [];

      exclude = Array.isArray(exclude) ? exclude : [exclude];
      exclude = Array.from(new Set([...EXCLUDE_BASE, ...exclude]));

      if (shouldScan) {
        initialPresets.push(...(await scan({ dir, exclude, presets })));
      }

      const choices = Array.from(presets.entries())
        .sort()
        .map(([name, item]) => {
          return {
            name: item.description,
            value: name,
            checked: initialPresets.includes(name),
          };
        });

      inquirer
        .prompt([
          {
            type: "checkbox",
            message: "Choose your project's presets",
            name: "presets",
            choices,
          },
        ])
        .then(async (answers: Answers) => {
          if (answers.presets) {
            const config = createConfig({ presets });
            const stringifiedConfig = JSON.stringify(config, null, "  ");
            const outputFile = createConfigFilename(dir);

            return writeFile(outputFile, stringifiedConfig);
          }
        });
    }
  );

cli.help();
cli.parse();
