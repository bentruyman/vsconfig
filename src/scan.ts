import klaw from "klaw";

import { relative } from "path";

import type { PresetMap } from "./presets";
import { findMatchingPresets } from "./util";

interface ScanOptions {
  dir: string;
  presets: PresetMap;
}

export function scan({ dir, presets }: ScanOptions): Promise<string[]> {
  return new Promise((resolve) => {
    const foundPresets: Set<string> = new Set();

    klaw(dir)
      .on("data", (item) => {
        const rootRelativeFilename = relative(dir, item.path);

        findMatchingPresets(rootRelativeFilename, presets).forEach((key) =>
          foundPresets.add(key)
        );
      })
      .on("end", () => {
        resolve(Array.from(foundPresets));
      });
  });
}
