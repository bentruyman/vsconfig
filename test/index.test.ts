import { createConfig } from "../src";

describe("createConfig", () => {
  it("creates a new config using the current folder as the project root by default", () => {
    const config = createConfig({
      presets: ["json", "markdown", "typescript"],
    });

    expect(config.folders[0]).toEqual({ path: "." });
  });

  it("creates a new config based on a collection of specified presets", () => {
    const config = createConfig({
      presets: ["json", "markdown", "typescript"],
    });

    expect(Object.keys(config.settings)).toHaveLength(3);
    expect(config.settings["[json]"]).toBeDefined();
    expect(config.settings["[markdown]"]).toBeDefined();
    expect(config.settings["[typescript]"]).toBeDefined();
    expect(config).toMatchSnapshot();
  });
});
