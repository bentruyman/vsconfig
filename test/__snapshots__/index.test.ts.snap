// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createConfig creates a new config based on a collection of specified presets 1`] = `
Object {
  "folders": Array [
    Object {
      "path": ".",
    },
  ],
  "settings": Object {
    "[json]": Object {
      "editor.defaultFormatter": "esbenp.prettier-vscode",
      "editor.formatOnSave": true,
    },
    "[markdown]": Object {
      "editor.defaultFormatter": "esbenp.prettier-vscode",
      "editor.formatOnSave": true,
    },
    "[typescript]": Object {
      "editor.defaultFormatter": "dbaeumer.vscode-eslint",
      "editor.formatOnSave": true,
    },
  },
}
`;
